### PR TITLE
support proxy_cache

### PR DESCRIPTION
**Pull request by kumagi**
When buffer is written in disk, we should read it.

### CODE REVIEW

File: ngx_http_sxg_filter_module.c [hunk 1]
```diff
@@ -441,12 +441,24 @@ static bool copy_buffer_to_sxg_buffer(const ngx_chain_t* in, sxg_buffer_t* buf,
...
+                          cl->buf->file_pos);
+        if (copied_size != copy_size) {
+          return false;
+        }
+      } else if (cl->buf->memory) {
+        const size_t copy_size = cl->buf->last - cl->buf->pos;
+        if (buf->size + copy_size > limit ||
+            !sxg_write_bytes(cl->buf->pos, copy_size, buf)) {
+          return false;
+        }
```
Comment (gaul) [inline ngx_http_sxg_filter_module.c]: Should this have an `else` clause returning `false` or emitting some warning?

File: ngx_http_sxg_filter_module.c [hunk 2]
```diff
@@ -441,12 +441,24 @@ static bool copy_buffer_to_sxg_buffer(const ngx_chain_t* in, sxg_buffer_t* buf,
...
-      const size_t copy_size = cl->buf->last - cl->buf->pos;
-      if (buf->size + copy_size > limit ||
-          !sxg_write_bytes(cl->buf->pos, copy_size, buf)) {
-        return false;
+      if (cl->buf->in_file) {
+        const size_t copy_size = cl->buf->file_last - cl->buf->file_pos;
+        const size_t buffer_tail = buf->size;
+        sxg_buffer_resize(buf->size + copy_size, buf);
+        const ssize_t copied_size =
+            ngx_read_file(cl->buf->file, &buf->data[buffer_tail], copy_size,
```
Comment (gaul) [inline ngx_http_sxg_filter_module.c]: Can just use `buf->data + buffer_tail`.

File: ngx_http_sxg_filter_module.c [hunk 3]
```diff
@@ -441,12 +441,24 @@ static bool copy_buffer_to_sxg_buffer(const ngx_chain_t* in, sxg_buffer_t* buf,
   *last_buf = false;
   for (const ngx_chain_t* cl = in; cl != NULL; cl = cl->next) {
     if (cl->buf != NULL) {
-      const size_t copy_size = cl->buf->last - cl->buf->pos;
-      if (buf->size + copy_size > limit ||
-          !sxg_write_bytes(cl->buf->pos, copy_size, buf)) {
-        return false;
+      if (cl->buf->in_file) {
```
Comment (gaul) [inline ngx_http_sxg_filter_module.c]: Should this take into account `limit`?

File: ngx_http_sxg_filter_module.c (hunk 2 above)
Comment (kumagi) [reply ngx_http_sxg_filter_module.c]: OK.

File: ngx_http_sxg_filter_module.c (hunk 3 above)
Comment (kumagi) [reply ngx_http_sxg_filter_module.c]: OK, I'll add a check.

File: ngx_http_sxg_filter_module.c (hunk 1 above)
Comment (kumagi) [reply ngx_http_sxg_filter_module.c]: OK, add error message.

Comment (gaul) [reply ngx_http_sxg_filter_module.c]: Sorry I meant `if (!cl->buf->in_file && !cl->buf->memory)` what should happen?  You could at least `return false`, possibly with a log message.

Comment (kumagi) [reply ngx_http_sxg_filter_module.c]: `cl` is a node of a linked list.
It contains buffer (maybe in disk or not), or signal (flush or EOF), so `!cl->buf->in_file && !cl->buf->memory` can happen for some reason. In that case we should ignore the node in this context I believe.

Comment (gaul) [reply ngx_http_sxg_filter_module.c]: Thanks for the explanation!

**Pull request merged**